gcloud-java-dns/src/main/java/com/google/gcloud/dns/DnsImpl.java

```diff
@@ -0,0 +1,335 @@
+      this.serviceOptions = serviceOptions;
+    }
+
+    @Override
+    public Page<ChangeRequest> nextPage() {
+      return listChangeRequests(zoneName, serviceOptions, requestOptions, PB_TO_CHANGE_REQUEST);
+    }
+  }
+
+  private static class DnsRecordPageFetcher implements PageImpl.NextPageFetcher<DnsRecord> {
+
+    private static final Function<ResourceRecordSet, DnsRecord> PB_TO_DNS_RECORD =
+        new Function<ResourceRecordSet, DnsRecord>() {
+          @Override
+          public DnsRecord apply(com.google.api.services.dns.model.ResourceRecordSet pb) {
+            return DnsRecord.fromPb(pb);
+          }
+        };
+    private static final long serialVersionUID = 670996349097667660L;
+    private final Map<DnsRpc.Option, ?> requestOptions;
+    private final DnsOptions serviceOptions;
+    private final String zoneName;
+
+    DnsRecordPageFetcher(String zoneName, DnsOptions serviceOptions, String cursor,
+        Map<DnsRpc.Option, ?> optionMap) {
+      this.zoneName = zoneName;
+      this.requestOptions =
+          PageImpl.nextRequestOptions(DnsRpc.Option.PAGE_TOKEN, cursor, optionMap);
+      this.serviceOptions = serviceOptions;
+    }
+
+    @Override
+    public Page<DnsRecord> nextPage() {
+      return listDnsRecords(zoneName, serviceOptions, requestOptions, PB_TO_DNS_RECORD);
+    }
+  }
+
+  private static Page<Zone> listZones(final DnsOptions serviceOptions,
+      final Map<DnsRpc.Option, ?> optionsMap) {
+    // define transformation function
+    // this differs from the other list operations since zone is functional and requires dns service
+    Function<ManagedZone, Zone> pbToZoneFunction = new Function<ManagedZone, Zone>() {
+      @Override
+      public Zone apply(
+          com.google.api.services.dns.model.ManagedZone zonePb) {
+        return new Zone(serviceOptions.service(), ZoneInfo.fromPb(zonePb));
+      }
+    };
+    try {
+      // get a list of managed zones
+      DnsRpc.ListResult<ManagedZone> result =
+          runWithRetries(new Callable<DnsRpc.ListResult<ManagedZone>>() {
+            @Override
+            public DnsRpc.ListResult<ManagedZone> call() {
+              return serviceOptions.rpc().listZones(optionsMap);
+            }
+          }, serviceOptions.retryParams(), EXCEPTION_HANDLER);
+      String cursor = result.pageToken();
+      // transform that list into zone objects
+      Iterable<Zone> zones = result.results() == null
+          ? ImmutableList.<Zone>of() : Iterables.transform(result.results(), pbToZoneFunction);
+      return new PageImpl<>(new ZonePageFetcher(serviceOptions, cursor, optionsMap),
+          cursor, zones);
+    } catch (RetryHelperException e) {
+      throw DnsException.translateAndThrow(e);
+    }
+  }
+
+  private static Page<ChangeRequest> listChangeRequests(final String zoneName,
+      final DnsOptions serviceOptions, final Map<DnsRpc.Option, ?> optionsMap,
+      Function<Change, ChangeRequest> TRANSFORM_FUNCTION) {
```

Comment:
if you think there is a value in passing TRANSFORM_FUNCTION to this method (as oppose to referring to it directly from the method) then change the param to a valid local-variable/param name.